### ISSUE TITLE
2.8.4 hang at avahi_register when Pi-3 + eth0 + wlan0 running in AP mode

### ISSUE THREAD

**Issue by moodeaudio** [mikebrady/shairport-sync#319]
Hi Mike,

Testing 2.8.4 on a Pi-3 with eth0 connected and wlan0 configured as Access Point (AP) and I'm getting hang at avahi_register. Ifconfig output and log trace below.

AP mode is something that Moode 2.6 does automatically for wlan0 if adapter exists and there is no SSID already specified. It enables direct WiFi connection to Moode using "Moode" SSID. AP mode provided by hostapd and dnsmasq and uses 172.24.1.1/24 addr range.

This particular issue would affect a couple of common Moode configs:
1) Pi-3 + eth0 connection
2) Pi-2 and earlier + eth0 connection and an AP mode capable WiFi adapter plugged in.

-Tim

IFCONFIG

pi@moode:~ $ ifconfig
eth0      Link encap:Ethernet  HWaddr b8:27:eb:cd:fc:4e  
          inet addr:192.168.1.213  Bcast:192.168.1.255  Mask:255.255.255.0
          inet6 addr: fe80::3f32:d353:e7b4:a6c9/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:1504 errors:0 dropped:0 overruns:0 frame:0
          TX packets:266 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:170854 (166.8 KiB)  TX bytes:57233 (55.8 KiB)

lo        Link encap:Local Loopback  
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:222 errors:0 dropped:0 overruns:0 frame:0
          TX packets:222 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1 
          RX bytes:105501 (103.0 KiB)  TX bytes:105501 (103.0 KiB)

wlan0     Link encap:Ethernet  HWaddr b8:27:eb:98:a9:1b  
          inet addr:172.24.1.1  Bcast:172.24.1.255  Mask:255.255.255.0
          inet6 addr: fe80::324c:cae4:5466:861a/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:109 errors:0 dropped:109 overruns:0 frame:0
          TX packets:60 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:18237 (17.8 KiB)  TX bytes:9335 (9.1 KiB)

LOG TRACE
pi@moode:~ $ /usr/local/bin/shairport-sync -vv -a "Moode Airplay" -S soxr -w -B /var/www/command/spspre.sh -E /var/www/command/spspost.sh --metadata-pipename=/tmp/shairport-sync-metadata --get-coverart -- -d hw:0 -c PCM
Looking for the configuration file "/etc/shairport-sync.conf".
Looking for configuration file at full path "/etc/shairport-sync.conf"
Output device name is "hw:0".
Open Mixer
Lowest dB value is a mute -- try minimum volume +1
Hardware mixer has dB volume from -102.380000 to 4.000000.
Has mute ability.
The processor is running little-endian.
Version: "2.8.4-OpenSSL-Avahi-ALSA-stdout-soxr-metadata"
statistics_requester status is 0.
daemon status is 0.
rtsp listening port is 5000.
udp base port is 6001.
udp port range is 100.
Shairport Sync player name is "Moode Airplay".
Audio Output name is "(null)".
on-start action is "/var/www/command/spspre.sh".
on-stop action is "/var/www/command/spspost.sh".
wait-cmd status is 1.
mdns backend "(null)".
userSuppliedLatency is 0.
AirPlayLatency is -1.
iTunesLatency is -1.
forkedDaapdLatency is -1.
stuffing option is "1" (0-basic, 1-soxr).
resync time is 2205.
allow a session to be interrupted: 0.
busy timeout time is 120.
drift tolerance is 88 frames.
password is "(null)".
ignore_volume_control is 0.
playback_mode is 0 (0-stereo, 1-mono).
disable_synchronization is 0.
audio backend desired buffer length is 6615.
audio backend latency offset is 0.
volume range in dB (zero means use the range specified by the mixer): 0.
zeroconf regtype is "_raop._tcp".
configuration file name "/etc/shairport-sync.conf" resolves to "/etc/shairport-sync.conf".
metdata enabled is 1.
metadata pipename is "/tmp/shairport-sync-metadata".
metadata socket address is "(null)" port 0.
metadata socket packet size is "500".
get-coverart is 1.
Successful Startup
avahi: avahi_register.

Log never gets beyond this point...

**Comment by moodeaudio** [comment]
Hi Mike,

I should add that a work around for the issue is to set a bogus SSID for wlan0 in Moode Network config screen. This causes Moode startup to bypass auto-AP mode because it sees a user-configured SSID for wlan0. Since its a bogus SSID, wlan0 does not get IP address. Ifconfig looks like below in this case. This is on Pi-3.

IFCONFIG

pi@moode:~ $ ifconfig
eth0      Link encap:Ethernet  HWaddr b8:27:eb:cd:fc:4e  
          inet addr:192.168.1.213  Bcast:192.168.1.255  Mask:255.255.255.0
          inet6 addr: fe80::3f32:d353:e7b4:a6c9/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:311 errors:0 dropped:0 overruns:0 frame:0
          TX packets:229 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:47297 (46.1 KiB)  TX bytes:59051 (57.6 KiB)

lo        Link encap:Local Loopback  
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:204 errors:0 dropped:0 overruns:0 frame:0
          TX packets:204 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1 
          RX bytes:103132 (100.7 KiB)  TX bytes:103132 (100.7 KiB)

wlan0     Link encap:Ethernet  HWaddr b8:27:eb:98:a9:1b  
          inet6 addr: fe80::324c:cae4:5466:861a/64 Scope:Link
          UP BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:3 errors:0 dropped:1 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:768 (768.0 B)  TX bytes:0 (0.0 B)

**Comment by mikebrady** [comment]
Hi Tim. Is this a new behaviour? When I use the Pi as a WiFi access point with `hostapd` I've always had to delay starting Shaiprprt Sync until after it starts. But I've only been trying to do this fairly recently.

**Comment by moodeaudio** [comment]
Hi Mike - After testing I see it's present in 2.8.3 as well. I never noticed it because I had always been testing shairport-sync on systems that were not running AP mode.

In my tests I was starting shairport-sync manually from Moode System config screen which would mean that hostapd would already have been running by then since it is started during initial Moode startup.

**Comment by mikebrady** [comment]
It really did frustrate me a few weeks ago when I was trying to install it in a car, and I kind-of concluded that it was an issue with Avahi. If it was something that happened with a recent version of SPS and wasn't there before, that would be offer a great clue.

**Comment by moodeaudio** [comment]
If u have an older binary, feel free to email it to me and I'l run some tests. OS I'm using is Jessie-Lite with 4.4.8 kernel.

**Comment by moodeaudio** [comment]
Hi Mike - the 2.6 binary that u sent behaves same so probably some quirk in avahi...

**Comment by mikebrady** [comment]
Thanks tim. Either an Avahi bug or a long-standing bug in Shairport Sync. It needs some more investigation.

**Comment by jobino** [comment]
how to solve this problem?

**Comment by mikebrady** [comment]
The workaround is to wait until the AP has been established and the network configuration is stable before starting Shairport Sync. That's what works for me.

**Comment by mikebrady** [comment]
There have been many updates since this, and I'm pretty sure the problem has been resolved. Please open a new issue if necessary.
